Update Dataclasses docs (#6470)
pydantic/pydantic: docs/migration.md

    * As a result, the `__post_init_post_parse__` method would have become redundant, so has been removed.
* Pydantic no longer supports `extra='allow'` for Pydantic dataclasses, where extra fields passed to the initializer would be
    stored as extra attributes on the dataclass. `extra='ignore'` is still supported for the purpose of ignoring
    unexpected fields while parsing data, they just won't be stored on the instance.
* Pydantic dataclasses no longer have an attribute `__pydantic_model__`, and no longer use an underlying `BaseModel`
    to perform validation or provide other functionality.
    * To perform validation, generate a JSON schema, or make use of
        any other functionality that may have required `__pydantic_model__` in V1, you should now wrap the dataclass
        with a [`TypeAdapter`](usage/models.md#typeadapter) ([discussed more below](#introduction-of-typeadapter)) and
        make use of its methods.
* In Pydantic V1, if you used a vanilla (i.e., non-Pydantic) dataclass as a field, the config of the parent type would
    be used as though it was the config for the dataclass itself as well. In Pydantic V2, this is no longer the case.

>     * In Pydantic V2, to override the config like you would with a `BaseModel`, you can use the `config` parameter

Author: ```suggestion
    * In Pydantic V2, to override the `config` (like you would with a `BaseModel`), you can use the `config` parameter
```

### Changes to dataclasses

> Pydantic [dataclasses](usage/dataclasses.md) continue to be useful for enabling the data validation on standard
> dataclasses without having to subclass `BaseModel`. However, we have made some changes to the behavior of Pydantic
> dataclasses in V2:

Reviewer: I haven't seen `we` used in the docs yet – perhaps change language to be more subjective.

```suggestion
Pydantic [dataclasses](usage/dataclasses.md) continue to be useful for enabling the data validation on standard
dataclasses without having to subclass `BaseModel`. Pydantic V2 introduces the following changes to this dataclass behavior:
```

Reviewer: Oh, lol, I do see `we` used... Welp.

* When used as fields, dataclasses (Pydantic or vanilla) no longer accept tuples as validation inputs; dicts should be
  used instead.
* The `__post_init__` in Pydantic dataclasses will now be called _after_ validation, rather than before.
    * As a result, the `__post_init_post_parse__` method would have become redundant, so has been removed.

> * We no longer support `extra='allow'` for Pydantic dataclasses, where extra fields passed to the initializer would be

Reviewer: Nit: remove `we` (I see it exists before your edits):
```
* Pydantic dataclasses longer support for `extra='allow'` where extra fields passed to the initializer would be
```

> * We no longer support `extra='allow'` for Pydantic dataclasses, where extra fields passed to the initializer would be

Author: ```suggestion
* Pydantic no longer supports `extra='allow'` for Pydantic dataclasses, where extra fields passed to the initializer would be
```

    stored as extra attributes on the dataclass. `extra='ignore'` is still supported for the purpose of ignoring
    unexpected fields while parsing data, they just won't be stored on the instance.
* Pydantic dataclasses no longer have an attribute `__pydantic_model__`, and no longer use an underlying `BaseModel`
    to perform validation or provide other functionality.
    * To perform validation, generate a JSON schema, or make use of
        any other functionality that may have required `__pydantic_model__` in V1, you should now wrap the dataclass
        with a [`TypeAdapter`](usage/models.md#typeadapter) ([discussed more below](#introduction-of-typeadapter)) and
        make use of its methods.
* In Pydantic V1, if you used a vanilla (i.e., non-Pydantic) dataclass as a field, the config of the parent type would
    be used as though it was the config for the dataclass itself as well. In Pydantic V2, this is no longer the case.
    * In Pydantic V2, to override the config like you would with a `BaseModel`, you can use the `config` parameter

>         on the `dataclass` decorator. See [Dataclass Config](usage/dataclasses.md#dataclass-config) for examples.

Reviewer: _(This one is personal style related - feel free to ignore)_
```
    * In Pydantic V2, to override the `config` (like you would with a `BaseModel`), you can use the `config` parameter
        on the `dataclass` decorator. See [Dataclass Config](usage/dataclasses.md#dataclass-config) for examples.
```

>         on the `dataclass` decorator. See [Dataclass Config](usage/dataclasses.md#dataclass-config) for examples.

Reviewer: Is it `config` or `model_config`?

Author: I believe it's the `config` option on `@dataclass` in this context.

Reviewer: Right you are. Worth mentioning that this is different from BaseModels? (Or is that info cramming?)

Author: @ybressler reworded based on your feedback, hopefully this is more clear about BaseModel vs dataclass config
